app-media-image-capture.d.ts

```diff
@@ -0,0 +1,158 @@
+
+  /**
+   * The last photo captured by the image capture instance.
+   */
+  readonly lastPhoto: Blob|null;
+
+  /**
+   * The last frame grabbed by the image capture instance.
+   */
+  readonly lastFrame: ImageBitmap|null;
+
+  /**
+   * The PhotoCapabilities for the device providing the image data
+   * associated with the chosen video track. This object contains
+   * information about the minimum, maximum, current and incremental
+   * values for various camera settings.
+   */
+  readonly photoCapabilities: object|null|undefined;
+
+  /**
+   * The MediaTrackCapabilities for the device providing the image data
+   * associated with the chosen video track. This object contains
+   * information about the minimum, maximum, current and incremental
+   * values for various camera settings.
+   */
+  readonly trackCapabilities: object|null|undefined;
+
+  /**
+   * The PhotoSettings that will be used to configure the ImageCapture
+   * instance used by this element. This configuration is generated
+   * based on the individually configured properties on this element.
+   * A full list of configurable properties can be found
+   * [here](https://w3c.github.io/mediacapture-image/##photosettings-section).
+   */
+  readonly photoSettings: PhotoSettings;
+  fillLightMode: Polymer.AppMedia.FillLightMode|null;
+  imageHeight: number|null|undefined;
+  imageWidth: number|null|undefined;
+
+  /**
+   * Note that red eye reduction may not be controllable. If it is reported as
+   * controllable, the value of this property will be respected. Otherwise it
+   * will be ignored.
+   */
+  redEyeReduction: boolean|null|undefined;
+
+  /**
+   * The constraints that will be applied to the MediaStreamTrack that
+   * is associated with the ImageCapture instance. This configuration is
+   * generated based on the individually configured properties on this
+   * element. A full list of configurable constraints can be found
+   * [here](https://w3c.github.io/mediacapture-image/#constrainable-properties)
+   *
+   * Note that if a given setting is not supported by the current track, it
+   * will be ignored. Also, any constraints that are suported will be clamped
+   * to the bounds that are reported by the track PhotoCapabilities instance.
+   */
+  readonly trackConstraints: MediaTrackConstraints;
+  whiteBalanceMode: string|null|undefined;
+  colorTemperature: number|null|undefined;
+  exposureMode: Polymer.AppMedia.MeteringMode|null;
+  exposureCompensation: number|null|undefined;
+  iso: number|null|undefined;
+  focusMode: Polymer.AppMedia.MeteringMode|null;
+  pointsOfInterest: any[]|null|undefined;
+  brightness: number|null|undefined;
+  contrast: number|null|undefined;
+  saturation: number|null|undefined;
+  sharpness: number|null|undefined;
+  zoom: number|null|undefined;
+  torch: boolean|null|undefined;
+
+  /**
+   * Take a photo. Returns a promise that resolves a photographic Blob.
+   *
+   * @returns Promise<Blob>
+   */
+  takePhoto(): any;
```

Comment:
This type seems wrong...

Reply:
Yup, should be `@return`

Reply:
I think both forms are accepted by closure compiler

Reply:
Yeah, both are valid. I used `@returns` in the generator because it seemed to be the canonical form for JSDoc (http://usejsdoc.org/tags-returns.html), although for Closure the canonical form is `@return` (https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler#return-type-description), and that's what we use in most of the time in the elements, so I now I'd prefer `@return`. But it didn't seem worth the churn of rewriting all these files over that.

Reply:
The actual problem here and below is that the source is missing curly braces around the type: `@return {Promise<Blob>}` (`@returns` would also work).

Reply:
Oh and... that would be `@return {!Promise<!Blob>}` because object types in Closure are nullable by default.

Reply:
Thanks for the clarification @aomarks 🍻

@TimvdLippe would you be willing to fix the source annotation as part of this change?

Reply:
Yes I will push an update tomorrow